### DISCUSSION TITLE
Changing cost function structure through opti.parameter

### DISCUSSION THREAD

**Discussion by itbellix** [casadi/casadi#3280] [Q&A] [answered]
Hi all,
I would like to update my cost function structure based on a parameter of the optimal control problem, when solving it in an receding horizon fashion (MPC). Abstracting a bit, I would like to consider multiple terms in the cost, but without knowing a priori how many they are at each iteration of the MPC. The only prior about these terms is that they are just variations of the same function, with different values of certain constants. 

I am using the Opti stack, defining an opti parameter related to the number of terms in the cost, and then I would ideally convert the optimization into a CasADi function, that I would like to call with a different value of the parameter.

I am unsure if CasADi supports my application, as I realize that changing the cost function "on the fly" is not ideal. To my understanding, redefining the cost function should in general be avoided as it changes the computational graph that CasADi leverages. However, since the cost terms I would add are just different "versions" of the same mathematical function, I would think that these updates could be done with minimal modifications of the underlying graph. I tried is to use the opti parameter in a for loop, so that I can use to add the various cost terms dynamically, but this strategy does not appear to work, as it requires to know the value of the opti parameter when the problem is being built. Below a sketch of my code, maybe clearer than my explanations:

```python
import casadi as ca

# I would like to have a cost function of the form J = J_1+J_2+..+J_K, 
# where each J_k differs from the others just from some parameter

num_parameters = 3          # number of parameters required to define each J_k

opti = ca.Opti()

K = opti.parameter(1)       # number of cost terms, that I would like to update once I convert the Opti object 
                            # to a CasADi function

parameters = opti.parameter(K*num_parameters)

J = 0                       # initialize cost function

# formulation of the NLP, inside which I would like to update the cost terms
for i in range(N):
    ...
    ...
    for term_k in range(K):
        # here I would iteratively add the contribution of each J_k to the cost, 
        # based on the parameters that pertain to each term  
    ...
    ...
    
# after building the NLP, I would transform it into a CasADi function 
# using something like MPC_iter = opti.to_function('MPC_iter', [K], [...])
```
The main problem that I encounter is that the last for loop cannot be executed as `K` is not known yet (I tried to set a dummy value for it using `opti.set_value()`, but this does not solve the issue). So my questions would be: 
1. is there a way to use a similar approach with some adjustments?
2. if not, is it just fundamentally wrong or are there other workarounds to achieve what I am trying to do in CasADi?

**Comment by brunoml5** [comment, marked as answer]
Hi!

What you need to do is write your objective with all the different functions. Then, use an `opti.parameter` multiplying each function to activate/deactivate them before each solver call. Something like this:

```
import casadi as ca

opti = ca.Opti()

x = opti.variable()
y = opti.variable()

p_J = opti.parameter(2)

J0 = (x-y**2)**2
J1 = (y-x**2)**2

J = p_J[0]*J0 + p_J[1]*J1

opti.minimize(J)

opti.subject_to(x**2 + y**2 == 1)
opti.subject_to(x + y >= 1)

opti.solver('ipopt')
```

Before each solve, set the parameters to 0s and 1s to activate the desired objective.

```
opti.set_value(p_J, [0, 1])
sol = opti.solve()

print(sol.value(x))
print(sol.value(y))
```

**Reply by itbellix** [reply]
Hi @brunoml5, thank you very much for your suggestion!
